lib/bundle/brew_installer.rb

```diff
@@ -24,14 +24,43 @@ def install_or_upgrade
       end
     end
 
-    def self.upgradable_formulae
-      outdated_formulae - pinned_formulae
+    def self.formula_installed_and_up_to_date?(formula)
+      formula_installed?(formula) && !formula_upgradable?(formula)
     end
 
     private
 
+    def self.formula_installed?(formula)
+      installed_formulae.include?(resolved_formula_name(formula))
+    end
+
+    def self.formula_upgradable?(formula)
+      upgradable_formulae.include?(resolved_formula_name(formula))
+    end
+
     def self.installed_formulae
-      @@installed_formulae ||= `brew list -1`.split("\n")
+      @@installed_formulae ||= Bundle::BrewDumper.new.formulae.map { |f| f[:name] }
```

Comment:
`Bundle::BrewDumper.new.formulae` this is probably what we want to cache to avoid duplicated formula topology sort.